### PR TITLE
CentOS 7 docker based image

### PR DESCRIPTION
**Pull request by mr-exz**
added Dockerfile with: 
- clear cenots/7
- dotnet core 2.2
- gitversion 4.0.0

### CODE REVIEW

**Comment by asbjornu** [comment]
Thanks for this, @mr-exz. Is it possible to test the image somehow?

**Comment by mr-exz** [comment]
> Thanks for this, @mr-exz. Is it possible to add an automatic test that is run by Travis or similar, somehow?

i will check how you did it, and will try fit my Dockerfile into build process

**Comment by mr-exz** [comment]
@asbjornu @arturcic pls check cahnges

File: src/Docker/linux/centos/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM centos:7
+LABEL maintainers="GitTools Maintainers"
+
+ENV DOTNET_VERSION='2.2'
```
Comment (arturcic) [inline src/Docker/linux/centos/Dockerfile]: Could you set the version to 2.1 to have the same version as for other docker images? I'll make some changes later to add support for building both 2.1 and 2.2 versions

File: build.cake [hunk 2]
```diff
@@ -622,6 +623,7 @@ Task("Publish-DockerHub")
     {
         DockerPush("linux", "dotnetcore", parameters);
         DockerPush("linux", "fullfx", parameters);
+        DockerPush("linux", "centos", parameters);
```
Comment (asbjornu) [inline build.cake]: Please excuse my ignorance, but wow is this going to affect the naming on [DockerHub](https://hub.docker.com/u/gittools)? I don't quite understand where the existing names repositories [`gitversion-fullfx`](https://hub.docker.com/r/gittools/gitversion-fullfx), [`gitversion-dotnetcore`](https://hub.docker.com/r/gittools/gitversion-dotnetcore), [`gitversion`](https://hub.docker.com/r/gittools/gitversion) and [`libgit2sharp-mono`](https://hub.docker.com/r/gittools/libgit2sharp-mono) come from and how they relate to these `DockerPush()` statements. Will `centos` become a new `tag` in the `gitversion-dotnetcore` repository?

Comment (arturcic) [reply build.cake]: > @asbjornu  [`gitversion`](https://hub.docker.com/r/gittools/gitversion) and [`libgit2sharp-mono`](https://hub.docker.com/r/gittools/libgit2sharp-mono) are already obsolete, only fullfx and dotnetcore are maintained. 

As for centos, that is right, the current changes will not create a tag under dotnetcore - good catch, we'll need to adjust the scripts to allow several distros of linux under dotnetcore and fullfx

Comment (arturcic) [reply build.cake]: this is the method where the tags are created https://github.com/GitTools/GitVersion/blob/master/build/utils.cake#L174

Comment (asbjornu) [reply build.cake]: I see. I find it a bit unfortunate that we have 4 different repositories for variations of the same executable. Can we get rid of every repository that isn't called simply `gitversion` and follow [NGINX](https://hub.docker.com/_/nginx?tab=tags)' naming scheme for labels to distinguish different versions of the image? That would give us one repository with tags following the scheme `{version}-{platform}-{framework}`, for example:

- `4.0.0-centos`
- `4.0.1-beta2-debian-fx`
- `4.0.0-beta15-windows-dotnetcore`

I know doing anything about this scheme is outside the scope of this PR, I just want to have a discussion on this so I know which options we have.

Comment (arturcic) [reply build.cake]: I suggest creating a separate issue, I can then implement it, and after that we can adjust this PR to the new scheme.

Comment (arturcic) [reply build.cake]: @asbjornu please create an issue with the proposed scheme and we can discuss it there

Comment (asbjornu) [reply build.cake]: Sure, I've created #1572 for further discussion.

**Comment by arturcic** [comment]
@asbjornu can we postpone merging this till I get #1572 implemented ?

**Comment by asbjornu** [comment]
@arturcic, yep, that's the plan. 😃

**Comment by arturcic** [comment]
@mr-exz the #1572  was implemented and merged, can you rebase on top of master and add the CentOS support? Do you think we can also add Fedora support?

**Comment by mr-exz** [comment]
> @mr-exz the #1572 was implemented and merged, can you rebase on top of master and add the CentOS support? Do you think we can also add Fedora support?

i think we can, i'll take a look

**Comment by arturcic** [comment]
@mr-exz , I think you should move the Dockerfile to the src/Docker/linux/centos/netcoreapp2.1 folder similar to debian version. This way we'll be able to add net core 2.2 version as well

**Comment by arturcic** [comment]
Looks great! Thank you for your contribution!

**Comment by mr-exz** [comment]
Not at all

**Comment by mr-exz** [comment]
@arturcic can you provide link on new centos7/fedora27 docker images link ?

**Comment by arturcic** [comment]
Yesterday when the PR was merged the Linux build has failed, some unit tests for some reason failed, now they work again. The tags can be found at https://hub.docker.com/r/gittools/gitversion/tags

**Comment by arturcic** [comment]
@mr-exz can you confirm that those docker images are valid and work as expected as I don't have time to test them these days?

**Comment by mr-exz** [comment]
i will, in process

**Comment by arturcic** [comment]
Thanks

**Comment by mr-exz** [comment]
For me working fine
`
# docker run --rm -v "$(pwd):/repo" gittools/gitversion:5.0.0-linux-centos7-netcoreapp2.1 /repo
# docker run --rm -v "$(pwd):/repo" gittools/gitversion:5.0.0-linux-fedora27-netcoreapp2.1 /repo
`

**Comment by arturcic** [comment]
great, thanks for the fedora image as well.

**Comment by mr-exz** [comment]
Not at all :)

**Comment by arturcic** [comment]
Btw, are you in a position to add a new PR for netcoreapp2.2? for linux images? That would be a great addition

**Comment by mr-exz** [comment]
ok will do, it shouldn't be so difficult

**Comment by asbjornu** [comment]
Hm, I just discovered that the tag `gittools/gitversion:5.0.0-linux-centos7-netcoreapp2.1` has the version number `5.0.0` which indicates it’s a stable version. Since it isn’t, that’s a problem, no? Can we get the pre-release label into there so we only tag stable releases as such?

**Comment by arturcic** [comment]
There is also a tag `gittools/gitversion:5.0.0-beta2-18-linux-centos7-netcoreapp2.1` that is the same as the one you mentioned. Basically on every new build we re-tag the 5.0.0-linux-centos7-netcoreapp2.1 to the newest version, as well as we tag with a pre-release tag. That means the pre-release tag will not change, but the the other tag gets updated with every new commit. (it's is similar to how latest tag is used in other projects, meaning it's always pointing to the latest, stable or not stable)

**Comment by arturcic** [comment]
@asbjornu, you can review this https://github.com/GitTools/GitVersion/blob/master/build/utils.cake#L163-L196

**Comment by arturcic** [comment]
And the versioning here https://github.com/GitTools/GitVersion/blob/master/build/version.cake#L23-L24

**Comment by mr-exz** [comment]
> Btw, are you in a position to add a new PR for netcoreapp2.2? for linux images? That would be a great addition

https://github.com/GitTools/GitVersion/pull/1633

**Comment by asbjornu** [comment]
@arturcic, thanks for the explanation. That sounds sensible to me. 👍

**Pull request merged**